bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java

```diff
@@ -77,27 +76,34 @@ public void processBindingConfiguration(String context, Item item, String bindin
 			throw new BindingConfigParseException("context is not permitted to be null for item "+item);
 		}
 
-		Set<Item> items = contextMap.get(context);
-		if (items==null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if (items == null) {
+				items = new HashSet<Item>();
+				contextMap.put(context, items);
+			}
+
+			items.add(item);
 		}
-			
-		items.add(item);
 	}
 
 	/**
 	 * {@inheritDoc}
 	 */
 	public void removeConfigurations(String context) {
-		Set<Item> items = contextMap.get(context);
-		if(items!=null) {
-			for(Item item : items) {
+		Set<Item> items = null;
+		synchronized (contextMap) {
+			items = contextMap.get(context);
+			if (items != null) {
+				contextMap.remove(context);
+			}
+		}
+		if (items != null) {
+			for (Item item : items) {
 				// we remove all binding configurations for all items
 				bindingConfigs.remove(item.getName());
 				notifyListeners(item);
 			}
-			contextMap.remove(context);
 		}
 	}
 	
```

Comment:
Some bindings' BindingProvider implementations have a `getItem` method that references the `contextMap` to find an item by name, but I don't know if this code is ever reached from `notifyListeners` in Core.  My opinion, subject to @teichsta or @kaikreuzer saying otherwise, is that my changes are more correct, in that during a `removeConfigurations` call, bindings ought not be trying to locate items by name, since the bindings are in the process of being torn down.